extensions/amp-ad/0.1/amp-ad-cross-domain-iframe-handler.js

```diff
@@ -106,78 +104,105 @@ export class AmpAdApiHandler {
       // respectively
-      this.adResponsePromise_ = listenForOncePromise(this.iframe_,
+      const responsePromise = listenForOncePromise(this.iframe,
         'bootstrap-loaded', this.is3p_);
-      listenForOncePromise(this.iframe_, 'no-content', this.is3p_)
-          .then(() => this.noContent_());
+      this.adResponsePromise_ =
+          timerFor(this.baseInstance_.win).timeoutPromise(TIMEOUT_VALUE,
+          responsePromise,
+          'timeout waiting for ad response');
+      listenForOncePromise(this.iframe, 'no-content', this.is3p_).then(() => {
+        this.noContent_();
+      });
     }
 
     if (!opt_defaultVisible) {
       // NOTE(tdrl,keithwrightbos): This will not work for A4A with an AMP
       // creative as it will not expect having to send the render-start message.
-      this.iframe_.style.visibility = 'hidden';
+      this.iframe.style.visibility = 'hidden';
     }
 
     this.unlisteners_.push(this.viewer_.onVisibilityChanged(() => {
       this.sendEmbedInfo_(this.baseInstance_.isInViewport());
     }));
 
-    this.element_.appendChild(this.iframe_);
-    return loadPromise(this.iframe_).then(() => {
-      return timerFor(this.baseInstance_.win).timeoutPromise(TIMEOUT_VALUE,
-          this.adResponsePromise_,
-          'timeout waiting for ad response').catch(e => {
-            this.noContent_();
-            user().warn('amp-ad', e);
-          }).then(() => {
-            //TODO: add performance reporting;
-            if (this.iframe_) {
-              this.iframe_.style.visibility = '';
-            }
-          });
+    this.element_.appendChild(this.iframe);
+    return loadPromise(this.iframe).then(() => {
+      return this.adResponsePromise_.catch(e => {
+        this.noContent_();
+        user().warn('amp-ad', e);
+      }).then(() => {
+        //TODO: add performance reporting;
+        if (this.iframe) {
+          this.iframe.style.visibility = '';
+        }
+      });
     });
   }
 
-  /** See BaseElement.  */
-  unlayoutCallback() {
+  /**
+   * Cleans up the listeners on the cross domain ad iframe.
+   * And free the iframe resource
+   * @param {boolean=} opt_keep
+   */
+  freeXDomainIframe(opt_keep) {
     this.cleanup_();
-    if (this.iframe_) {
-      removeElement(this.iframe_);
-      this.iframe_ = null;
+    // If ask to keep the iframe.
+    // Use in the case of no-content and iframe is a master iframe.
+    if (opt_keep) {
+      return;
+    }
+    if (this.iframe) {
+      removeElement(this.iframe);
+      this.iframe = null;
     }
   }
 
   /**
-   * Cleans up listeners on the ad, and calls the no content callback, if one
-   * was provided.
+   * Update size based on render-start returned size.
+   */
+  renderStart_(info) {
+    const data = info.data;
+    if (this.uiHandler_) {
+      this.uiHandler_.displayRenderStartUI();
+    }
+    this.updateSize_(data.height, data.width,
+        info.source, info.origin);
+    if (this.baseInstance_.lifecyleReporter_) {
```

Comment:
get type check error here
>>>ERROR - Property lifecyleReporter_ never defined on AMP.BaseElement
    if (this.baseInstance_.lifecyleReporter_) {
Should I pass the into the handler as `lifecyleReporterHandler`?